api/scanner/periodic_scanner/periodic_scanner.go

```diff
@@ -1,26 +1,39 @@
 package periodic_scanner
 
 import (
-	"log"
+	"fmt"
 	"sync"
 	"time"
 
 	"github.com/photoview/photoview/api/graphql/models"
+	"github.com/photoview/photoview/api/log"
 	"github.com/photoview/photoview/api/scanner/scanner_queue"
 	"gorm.io/gorm"
 )
 
+type ScannerQueue interface {
+	AddAllToQueue() error
+}
+
+type RealScannerQueue struct{}
+
+func (r *RealScannerQueue) AddAllToQueue() error {
+	return scanner_queue.AddAllToQueue()
+}
+
 type periodicScanner struct {
 	ticker         *time.Ticker
 	ticker_changed chan bool
+	done           chan struct{}
 	mutex          *sync.Mutex
 	db             *gorm.DB
+	scannerQueue   ScannerQueue
 }
 
+var mainPeriodicScannerMutex sync.Mutex
```

Comment:
I know it's the easy way to fix race condition. In my opinion, I don't like a global locker. It's easy to lock with IO actions, which causes deadlocking or inter-locking when IO can't finish on time.

How urgent do you think to finish this problem? I'll cover this with #1208.

Reply:
From the code, we could remove this mutex. The `mainPeriodicScanner` should be created when the server launches, and it should not be changed during running. Of course the content of `mainPeriodicScanner` may be changed, and that should be protected by the inner mutex.